### ISSUE TITLE
TiDE

### ISSUE THREAD

**Issue by alexcolpitts96** [unit8co/darts#1726]
Google recently proposed a new architecture: TiDE (time-series dense encoder).

Paper: https://arxiv.org/abs/2304.08424
Blog: https://ai.googleblog.com/2023/04/recent-advances-in-deep-long-horizon.html

I implemented it; however, before documenting and adding tests I wanted somebody to sanity check my work.

**Comment by joshua-xia** [comment]
What is the progress of this model? I look forward to using it, Thanks!

**Comment by alexcolpitts96** [comment]
@joshua-xia you have great timing. 

I am actually working on this right now. Hopefully should be done today/tomorrow!

**Comment by joshua-xia** [comment]
> @joshua-xia you have great timing.
> 
> I am actually working on this right now. Hopefully should be done today/tomorrow!

Thanks your quickly response, google reseach has committed the model last month:
https://github.com/google-research/google-research/tree/master/tide
You may refer to it. 

Thanks again!

**Comment by alexcolpitts96** [comment]
That is perfect! The diagrams in the paper and architecture were really vague. This will clear up those things and likely fix and performance issues I introduced.

It might take a little longer, but the outcome should be better.

**Comment by joshua-xia** [comment]
@alexcolpitts96 

I run your implementation, the train and automatically save the checkpoint / best model has no problem.
But I did not know how to map your model parameter to the google official implememtation in the google reseach repository.
and the results (mse, mae etc) is not as good as google official implememtation based on the same dataset.

I think it is better to reproduce the paper's result base on your implementation,

The following is my test code:


def my_idx(idx):
    return (idx.year - 1950) / 50

# build the TiDEModel model
model = TiDEModel(
    input_chunk_length=128,
    output_chunk_length=48,
    num_encoder_layers=2,
    num_decoder_layers=2,
    decoder_output_dim=32,
    hidden_size=512,
    temporal_width=32,
    temporal_decoder_hidden=16,
    dropout=0.5,
    use_static_covariates=True,
    batch_size=512,
    n_epochs=200,
    nr_epochs_val_period=1,
    model_name="TiDEModel",
    force_reset=True,
    save_checkpoints=True,
    add_encoders={
        'cyclic': {'future': ['month']},
        'datetime_attribute': {'future': ['dayofweek']},
        'position': {'past': ['relative'], 'future': ['relative']},
        'custom': {'past': [my_idx]},
        'transformer': Scaler(StandardScaler())
    }
)

results = model.lr_find(
    series=train_1,
    val_series=val_1,
    past_covariates=train_1_covariates,
    val_past_covariates=val_1_covariates,
)

print(results.suggestion())

early_stopper = EarlyStopping("val_loss", min_delta=0.00001, patience=100, verbose=True)

model = TiDEModel(
    input_chunk_length=128,
    output_chunk_length=48,
    num_encoder_layers=2,
    num_decoder_layers=2,
    decoder_output_dim=32,
    hidden_size=512,
    temporal_width=32,
    temporal_decoder_hidden=16,
    dropout=0.5,
    use_static_covariates=True,
    batch_size=512,
    n_epochs=500,
    nr_epochs_val_period=1,
    model_name="TiDEModel",
    force_reset=True,
    save_checkpoints=True,
    optimizer_kwargs={"lr": results.suggestion()},
    pl_trainer_kwargs={"callbacks": [early_stopper]},
    add_encoders={
        'cyclic': {'future': ['month']},
        'datetime_attribute': {'future': ['dayofweek']},
        'position': {'past': ['relative'], 'future': ['relative']},
        'custom': {'past': [my_idx]},
        'transformer': Scaler(StandardScaler())
    }
)

model.fit(
    series=train_1,
    val_series=val_1,
    past_covariates=train_1_covariates,
    val_past_covariates=val_1_covariates,
)

model = TiDEModel.load_from_checkpoint("TiDEModel", best=True)
preds = model.predict(series=train_1, past_covariates=cov1_scaled, n=48)
mae_value = mae(val_1, preds)


I can help you on the test, feel free to let me know.

**Comment by alexcolpitts96** [comment]
I made all the changes requested by the devs, I am just waiting for them to check and merge.

I appreciate the offer though!

**Comment by madtoinou** [comment]
Implemented by #1727.
